### PR TITLE
Feature: namespace BQ metadata field when provided

### PR DESCRIPTION
**Pull request by rootcss**
when `BQ_METADATA_NAMESPACE=` config is set to some value, namespace the kafka offset metadata in BQ i.e, as a RECORD/STRUCT type.

By default, the value will be empty, which means metadata fields will be written as different columns rather than a RECORD/STRUCT type.

### CODE REVIEW

File: src/main/java/com/gojek/beast/protomapping/ProtoUpdateListener.java [hunk 1]
```diff
@@ -108,14 +108,28 @@ public void onProtoUpdate(String url, Map<String, DescriptorAndTypeName> newDesc
...
+        addMetadataFields(bqSchemaFields);
         bqClient.upsertTable(bqSchemaFields);
         protoMappingConfig.setProperty("PROTO_COLUMN_MAPPING", protoMappingString);
         setProtoParser(protoMappingConfig.getProtoColumnMapping());
     }
 
+    private void addMetadataFields(List<Field> bqSchemaFields) {
+        if (appConfig.getBqMetadataNamespace().isEmpty()) {
+            List<Field> bqMetadataFields = BQField.getMetadataFields();
+            bqSchemaFields.addAll(bqMetadataFields);
```
Comment (rootcss) [inline src/main/java/com/gojek/beast/protomapping/ProtoUpdateListener.java:120]: check for collision with all the metadata fields

**Comment by rootcss** [comment]
@sravankorumilli Please review and merge.

**Review by mauliksoneji — CHANGES_REQUESTED** [review]

File: src/test/java/com/gojek/beast/converter/ConsumerRecordConverterIntegrationTest.java [hunk 2]
```diff
@@ -158,6 +158,57 @@ public void shouldThrowExceptionForNullRecords() throws InvalidProtocolBufferExc
...
+        record1ExpectedColumns.put("bq_order_url", "order-url-1");
+        record1ExpectedColumns.put("bq_order_details", "order-details-1");
+        record1ExpectedColumns.putAll(util.metadataColumns(record1Offset, nowEpochMillis));
+
+        List<ConsumerRecord<byte[], byte[]>> messages = Arrays.asList(record1);
+        List<Record> records = recordConverter.convert(messages);
+
+        assertEquals(messages.size(), records.size());
+        Map<String, Object> record1Columns = records.get(0).getColumns();
+        assertEquals(record1ExpectedColumns.size(), record1Columns.size());
```
[the remark this answers is not in the document]
Comment (rootcss) [reply src/test/java/com/gojek/beast/converter/ConsumerRecordConverterIntegrationTest.java:180]: Oops, Removed it.

File: src/test/java/com/gojek/beast/converter/ConsumerRecordConverterIntegrationTest.java [hunk 3]
```diff
@@ -158,6 +158,57 @@ public void shouldThrowExceptionForNullRecords() throws InvalidProtocolBufferExc
...
+        record1ExpectedColumns.put("bq_order_url", "order-url-1");
+        record1ExpectedColumns.put("bq_order_details", "order-details-1");
+        record1ExpectedColumns.put(appConfig.getBqMetadataNamespace(), util.metadataColumns(record1Offset, nowEpochMillis));
+
+        List<ConsumerRecord<byte[], byte[]>> messages = Arrays.asList(record1);
+        List<Record> records = recordConverter.convert(messages);
+
+        assertEquals(messages.size(), records.size());
+        Map<String, Object> record1Columns = records.get(0).getColumns();
+        assertEquals(record1ExpectedColumns.size(), record1Columns.size());
```
[the remark this answers is not in the document]
Comment (rootcss) [reply src/test/java/com/gojek/beast/converter/ConsumerRecordConverterIntegrationTest.java:205]: Oops, Removed it.

**Review by mauliksoneji — APPROVED** [review]

**Pull request merged**